File: Python/Word Ladder.py
class Solution:
    # @param beginWord, a string
    # @param endWord, a string
    # @param wordDict, a set<string>
    # @return an integer
    def ladderLength(self, start, end, dict):
        set1 = set([start])
        set2 = set([end])
        layer = 0
        while len(set1):
            set3 = set([])
            for word in set1:
                for i in range(len(word)):
                    for ch in range(ord('a'), ord('z') + 1):
                        if word[i] == chr(ch):
                            continue
                        newword = word[:i] + chr(ch) + word[i+1:]
                        if newword in set2:
                            return layer + 2
                        if newword in dict:
                            set3.add(newword)
                            dict.remove(newword)
            if len(set3) <= len(set2):
                set1 = set3
            else:
                set1 = set2
                set2 = set3
            layer += 1
        return 0

File: java/Word_Ladder.java
// BFS soluton
// Time is O(n), space is O(n);

public class Solution {
    public int ladderLength(String beginWord, String endWord, Set<String> wordDict) {
        Queue<String> que = new LinkedList<String>();
        wordDict.add(beginWord);
        wordDict.add(endWord);
        int ret = 1;
        que.offer(beginWord);
        wordDict.remove(beginWord);
        while (que.peek() != null) {
            ret++;
            for (int n = que.size(); n >= 1; --n) {
                String s = que.poll();
                StringBuilder sb = new StringBuilder(s);
                for (int i = 0; i < s.length(); ++i) {
                    char pre = s.charAt(i);
                    for (char c = 'a'; c <= 'z'; ++c) {
                        if (c == pre ) continue;
                        sb.setCharAt(i,c);
                        String ss = sb.toString();
                        if (wordDict.contains(ss)) {
                            if (ss.equals(endWord)) 
                                return ret;
                            que.offer(ss);
                            wordDict.remove(ss);
                        }
                    }
                    sb.setCharAt(i,pre);
                }
            }
        }
        return 0;
    }
}
